tests/test_getTemplate.py

```diff
@@ -0,0 +1,265 @@
+                for p in points:
+                    display.dot(patchId, p.x, p.y, ctype=color)
+
+            # This is mostly stolen from pipe_tasks warpAndPsfMatch, but
+            # ip_diffim cannot depend on pipe_tasks.
+            xyTransform = lsst.afw.geom.makeWcsPairTransform(self.exposure.wcs, patch.wcs)
+            warpedPsf = lsst.meas.algorithms.WarpedPsf(self.exposure.psf, xyTransform)
+            warped = warper.warpExposure(patch.wcs, self.exposure, destBBox=box)
+            warped.setPsf(warpedPsf)
+            self.patches[tract.tract_id].append(warped)
+            self.dataIds[tract.tract_id].append({"tract": tract.tract_id,
+                                                 "patch": patchId,
+                                                 "band": "a"})
+
+    def _checkMetadata(self, template, config, box, wcs, nInputs):
+        """Check that the various metadata components were set correctly.
+        """
+        expectedBox = lsst.geom.Box2I(box)
+        expectedBox.grow(config.templateBorderSize)
+        self.assertEqual(template.getBBox(), expectedBox)
+        # WCS should match our exposure, not any of the coadd tracts.
+        for tract in self.patches:
+            self.assertNotEqual(template.wcs, self.patches[tract][0].wcs)
+        self.assertEqual(template.wcs, self.exposure.wcs)
+        self.assertEqual(template.photoCalib, self.exposure.photoCalib)
+        self.assertEqual(template.getXY0(), expectedBox.getMin())
+        self.assertEqual(template.filter.bandLabel, "a")
+        self.assertEqual(template.filter.physicalLabel, "a_test")
+        self.assertEqual(template.psf.getComponentCount(), nInputs)
+        # TODO: need other things to test here?
+
+    def _checkPixels(self, template, config, box):
+        """Check that the pixel values in the template are close to the
+        original image.
+        """
+        # All pixels should have real values!
+        expectedBox = lsst.geom.Box2I(box)
+        expectedBox.grow(config.templateBorderSize)
+
+        if debug:
+            _showTemplate(expectedBox, template)
+
+        # Check that we fully filled the template from the patches.
+        self.assertTrue(np.all(np.isfinite(template.image.array)))
+        # Because of the scale changes, there will be some ringing in the
+        # difference between the template and the original image; pick
+        # tolerances large enough to account for that.
+        self.assertImagesAlmostEqual(template.image, self.exposure[expectedBox].image,
+                                     rtol=.1, atol=4)
+        # Variance plane ==2 in the original image, but the warped images will
+        # have some structure due to the warping.
+        self.assertImagesAlmostEqual(template.variance, self.exposure[expectedBox].variance,
+                                     rtol=0.5, msg="variance planes differ")
```

Comment:
FWIW, if you want to be able to shrink this tolerance, I'd _guess_ the best way to do it would be to shrink the scaling from 1.05 to 1.01 or something.

Reply:
Unfortunately, that doesn't seem to change it much. The 3 degree rotation seems to contribute quite a bit.